File: Project/BTKInjectorTests/BTKInjectorDuckTypingTests.m
//
//  BTKInjectorDuckTypingTests.m
//  BTKInjector
//
//  Created by Tomohisa Ota on 2014/04/25.
//  Copyright (c) 2014 Tomohisa Ota. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "BTKInjector.h"
#import "BTKTestProtocol.h"
#import "BTKTestProtocol1Impl.h"
#import "BTKInjectorProxyProtocolOverride.h"

@interface BTKInjectorDuckTypingTests : XCTestCase

@end

@implementation BTKInjectorDuckTypingTests

- (void)testWithoutDuckTyping
{
    @try {
        id<BTKInjector> injector = [BTKGlobalInjector injectorWithBlock:^(id<BTKMutableInjector> mInjector) {
           [mInjector bindProtocol:@protocol(BTKTestProtocol4)
                   toProviderBlock:^id(id<BTKInjector> injector) {
                       return [BTKTestProtocol1Impl new];
                   }];
        }];
        id<BTKTestProtocol4> obj = [injector proxyFor:@protocol(BTKTestProtocol4)];
        [obj test1];
    }
    @catch (NSException *exception) {
        XCTAssertTrue(([exception.reason rangeOfString:@"does not conforms"].location != NSNotFound) ,
                      @"Exception type is not correct %@",exception);
        return;
    }
    XCTFail(@"Protocol mismatch did not detected");
}

- (void)testWithDuckTyping
{
    id<BTKInjector> injector = [BTKGlobalInjector injectorWithBlock:^(id<BTKMutableInjector> mInjector) {
        [mInjector bindProtocol:@protocol(BTKTestProtocol4)
                   forceConform:YES
                toProviderBlock:^id(id<BTKInjector> injector) {
                    return [BTKTestProtocol1Impl new];
                }];
    }];
    [injector proxyFor:@protocol(BTKTestProtocol4)];
}

@end

File: Project/BTKInjectorTests/Swift/BTKInjectorDuckTypingTests.swift
//
//  BTKInjectorDuckTypingTests.swift
//  BTKInjector
//
//  Created by mono on 1/14/15.
//  Copyright (c) 2015 Tomohisa Ota. All rights reserved.
//

import XCTest

/**
*  Swiftでは、XCTAssertThrowsSpecificNamed相当のassertionや、try-catchが無いのでtestWithoutDuckTypingが記述出来ない
*/
class BTKInjectorDuckTypingTests: XCTestCase {

    func testWithDuckTyping() {
        let injector = BTKGlobalInjector.injectorWithBlock { mInjector in
            mInjector.bindProtocol(BTKTestProtocol4.self, forceConform: true) { injector in
                return BTKTestProtocol1Impl()
            }
        }
        let proxy = injector.proxyFor(BTKTestProtocol4.self) as BTKTestProtocol1
        XCTAssertTrue(proxy.isKindOfClass(BTKTestProtocol1Impl.self))
        let obj = proxy as BTKTestProtocol1Impl
        XCTAssertEqual(obj.test1(), "test1")
    }
}
